lib/s3.py

```diff
@@ -4,8 +4,11 @@
 
 # Adapted from examples in
 # https://s3.amazonaws.com/doc/s3-developer-guide/RESTAuthentication.html
+# and
+# https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html
```

Comment:
This is all rather involved, and took you quite some time to figure out. Did you look into existing libraries like [s3](https://pypi.org/project/s3/) for that?

Reply:
In the end I ended up adding some debugging output to s3cmd to get some hints about what I was doing wrong, just to see how the hashed values were ending up different.  That was already at the "minor adjustments" stage, though, and ended up finding some really really dumb problems (like I was spelling "linodeobjets.com" instead of "linodeobjects.com").